doc/cimple.md

```diff
@@ -0,0 +1,452 @@
+turn `int`s into `enum`s. Type-cast is not permitted, because some values of
+type `int` are not in the enumeration.
+
+## `-Wenum-names`
+
+Checks that `enum` value constants have the same prefix as the `enum` type,
+except they should be SCREAMING_CASE instead of Camel_Snake. There are currently
+37 exemptions to this rule. New enums should follow the naming convention.
+
+**Reason:** this naming convention helps identify the type of an `enum` constant
+at first glance.
+
+## `-Wenum-to-string` (global)
+
+Checks that `_to_string` functions for `enum`s are complete.
+
+**Reason:** we provide `to_string` functions for `enum` but don't want to
+manually maintain them. This linter checks that the function is exactly what
+we want it to be, and the error message will say what the function should look
+like.
+
+## `-Wenum-unpack` (global)
+
+Checks that `_unpack` functions for `enum`s are complete.
+
+**Reason:** we provide `unpack` functions for `enum` but don't want to
+manually maintain them. This linter checks that the function is exactly what
+we want it to be, and the error message will say what the function should look
+like.
+
+## `-Wfunc-prototypes`
+
+Checks that empty parameter lists in C functions are written as `(void)`.
+
+**Reason:** old-style empty parameter lists written as `()` are risky, because
+C interprets them as variadic. GCC warns about this but sometimes misses one.
+
+## `-Wfunc-scopes`
+
+Checks that static function definitions are marked with `static`.
+
+In C, a function is `static` even if the definition doesn't use `static`, but
+there happens to be another declaration of the function which does.
+
+**Reason:** static/extern qualification of functions should be visible locally.
+It takes mental effort otherwise to look up the declaration to check for storage
+qualifiers.
+
+## `-Wglobal-funcs`
+
+Checks that no extern functions are declared in .c files.
+
+Extern functions must only be declared in .h files. In .c files all declarations
+must be static.
+
+**Reason:** extern declarations in .c files mean that we depend on a function
+not declared in a .h file we can include. This means we're depending on an
+unexported implementation detail, and there is no compiler that can check
+whether our declaration matches the implementation's definition.
+
+## `-Wlarge-struct-params`
+
+Checks that large structs are passed by pointer rather than by value.
+
+Exemptions are enums and some well-known small structs:
+
+- `Family`
+- `IP4`
+- `Logger_Level`
+- `Net_Packet_Type`
+- `Onion_Connection_Status`
+- `Packet`
+- `Packet_Direction`
+- `Socket`
+- `State_Type`
+
+and anything with one of the following prefixes, which are probably enums:
+
+- `Group_`
+- `MSI`
+- `Tox_`
+- `Toxav_`
+
+**Reason:** some structs in toxcore are up to 5MB in size, which would cause
+stack overflows. Since we can't currently measure the size, we avoid any struct
+passing altogether apart from some well-known exemptions.
+
+## `-Wlogger-calls`
```

Comment:
-Wformat-nonliteral could probably do this too?

Reply:
Not quite. We actually want it to be a true string literal, not a thing composed of macros and strings and not a bunch of string literals concatenated.